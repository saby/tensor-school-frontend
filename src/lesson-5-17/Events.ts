/*
   1. Создайте функцию createButton(). Необходимо, чтобы эта функция осуществила вставку в body тег button с текстом: "Удали меня".
      При клике по button удалить этот button.
*/
export function createButton(): void {
   document.body.insertAdjacentHTML('afterbegin', '<button>Удали меня</button>');
   document.body.getElementsByTagName('button')[0].addEventListener('click', (e) => {
      e.target.remove();
   });
}

/*
   2. Создайте функцию createArrList(arr), в которую передается 1 параметр: arr - массив строк.
      Функция выводит этот массив в виде маркированного списка внутри тега body.
      При наведении курсора мыши на элемент списка у этого элемента создается атрибут title, в котором записан его текст.
*/
export function createArrList(arr: string[]): void {
   let ul = document.createElement('ul');
   document.body.appendChild(ul);

   for (let i = 0; i < arr.length; i++) {
      let li = document.createElement('li');
      li.textContent = arr[i];
      li.addEventListener('mouseover', () => li.setAttribute('title', arr[i]))
      ul.appendChild(li);
   }
}

/*
   3. Создайте функцию createLink(), которая сгенерирует следующую разметку и вставит ее в body:

      <a href="https://tensor.ru/">tensor</a>

      При первом клике по ссылке в конец ее текста через пробел дописывается ее href.
      При следующем клике происходит действие по умолчанию (переход по ссылке в текущей вкладке).
*/
export function createLink(): void {
   let counter = 0;
   document.body.insertAdjacentHTML('afterbegin', '<a href="https://tensor.ru/">tensor</a>');
   document.body.getElementsByTagName('a')[0].addEventListener('click', (e) => {
      if (counter === 0) {
         counter++;
         e.preventDefault();
         e.target.innerHTML = e.target.innerHTML + ' ' + e.target.href;
      }
   });
}

/*
   4. Создайте функцию createList(), которая сгенерирует следующую разметку и вставит ее в body:

      <ul>
         <li>Пункт</li>
      </ul>
      <button>Добавить пункт</button>

      При клике по элементу li ему в конец текста добавляется восклицательный знак.
      При клике по button в конец списка добавляется новый элемент li с текстом: "Пункт".
      Клик по новому li также добавляет восклицательный знак в конец текста.
*/
export function createList(): void {
   let ul = document.createElement('ul');
   let button = document.createElement('button');
   button.textContent = 'Добавить пункт';

   document.body.appendChild(ul);
   document.body.appendChild(button);

   const addLi = function () {
      ul.insertAdjacentHTML('beforeend', '<li>Пункт</li>');
   }
   addLi();

   button.addEventListener('click', addLi);
   ul.addEventListener('click', (e) => {
       if (e.target && e.target.nodeName === 'LI') {
          e.target.innerHTML = e.target.innerHTML + '!';
       }
   });
}
