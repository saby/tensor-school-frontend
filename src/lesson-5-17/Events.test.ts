import {createButton, createArrList, createLink, createList} from './Events';

describe('Урок 5.17 - События', () => {
   describe('1. Создание и удаление кнопки по клику', () => {
      afterEach(() => {
         document.body.innerHTML = '';
      });

      it('Создана кнопка в body "Удали меня"', () => {
         createButton();
         expect(document.body).toMatchSnapshot();
      });

      it('Клик по кнопке "Удали меня"', () => {
         createButton();
         document.body.getElementsByTagName('button')[0].click();
         expect(document.body).toMatchSnapshot();
      });
   });

   describe('2. Создание списка и запись атрибута по наведению', () => {
      afterEach(() => {
         document.body.innerHTML = '';
      });

      it('Создан список в body из элементов массива', () => {
         createArrList(['нулевой элемент', 'первый элемент', 'второй элемент']);
         expect(document.body).toMatchSnapshot();
      });

      it('Наведение мыши на элементы списка', () => {
         const arr = ['нулевой элемент', 'первый элемент', 'второй элемент'];
         createArrList(arr);

         function triggerEvent(el: HTMLElement, type: string): void {
            if ('createEvent' in document) {
               var e = document.createEvent('HTMLEvents');
               e.initEvent(type, false, true);
               el.dispatchEvent(e);
            }
         }

         for (let i = 0; i < arr.length; i++) {
            triggerEvent(document.body.getElementsByTagName('li')[i], 'mouseover');
         }

         expect(document.body).toMatchSnapshot();
      });
   });

   describe('3. Создание ссылки и кликов по ней', () => {
      afterEach(() => {
         document.body.innerHTML = '';
      });

      it('Создана ссылка "tensor" в body', () => {
         createLink();
         expect(document.body).toMatchSnapshot();
      });

      it('Первый клик по ссылке "tensor"', () => {
         createLink();
         document.body.getElementsByTagName('a')[0].click();
         expect(document.body).toMatchSnapshot();
      });

      it('Второй клик по ссылке "tensor"', () => {
         createLink();

         interface ITestObj {
            testDefaultPrevented: boolean;
            testHref: string;
         }

         let testCounter = 0;
         let testObj: ITestObj = {
            testDefaultPrevented: false,
            testHref: ''
         };

         document.body.getElementsByTagName('a')[0].addEventListener('click', (e) => {
            testCounter++;

            if (testCounter < 2) {
               return;
            }

            // Второй клик
            testObj.testDefaultPrevented = e.defaultPrevented;
            testObj.testHref = e.target && e.target.getAttribute('href');

            expect(testObj).toEqual({
               testDefaultPrevented: false,
               testHref: 'https://tensor.ru/',
            });
         })

         document.body.getElementsByTagName('a')[0].click();
         document.body.getElementsByTagName('a')[0].click();
      });
   });

   describe('4. Создание списка, кнопки и кликов по ним', () => {
      afterEach(() => {
         document.body.innerHTML = '';
      });

      it('Создан список и кнопка', () => {
         createList();
         expect(document.body).toMatchSnapshot();
      });

      it('Клик по элементу списка', () => {
         createList();
         document.body.getElementsByTagName('li')[0].click();
         expect(document.body).toMatchSnapshot();
      });

      it('Клик по кнопке', () => {
         createList();
         document.body.getElementsByTagName('button')[0].click();
         expect(document.body).toMatchSnapshot();
      });

      it('Клик по созданному по кнопке элементу списка', () => {
         createList();
         document.body.getElementsByTagName('button')[0].click();
         document.body.getElementsByTagName('li')[1].click();
         expect(document.body).toMatchSnapshot();
      });
   });
});
