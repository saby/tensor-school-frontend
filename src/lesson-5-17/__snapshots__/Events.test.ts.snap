// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Урок 5.17 - События 1. Создание и удаление кнопки по клику Клик по кнопке "Удали меня" 1`] = `<body />`;

exports[`Урок 5.17 - События 1. Создание и удаление кнопки по клику Создана кнопка в body "Удали меня" 1`] = `
<body>
  <button>
    Удали меня
  </button>
</body>
`;

exports[`Урок 5.17 - События 2. Создание списка и запись атрибута по наведению Наведение мыши на элементы списка 1`] = `
<body>
  <ul>
    <li
      title="нулевой элемент"
    >
      нулевой элемент
    </li>
    <li
      title="первый элемент"
    >
      первый элемент
    </li>
    <li
      title="второй элемент"
    >
      второй элемент
    </li>
  </ul>
</body>
`;

exports[`Урок 5.17 - События 2. Создание списка и запись атрибута по наведению Создан список в body из элементов массива 1`] = `
<body>
  <ul>
    <li>
      нулевой элемент
    </li>
    <li>
      первый элемент
    </li>
    <li>
      второй элемент
    </li>
  </ul>
</body>
`;

exports[`Урок 5.17 - События 3. Создание ссылки и кликов по ней Первый клик по ссылке "tensor" 1`] = `
<body>
  <a
    href="https://tensor.ru/"
  >
    tensor https://tensor.ru/
  </a>
</body>
`;

exports[`Урок 5.17 - События 3. Создание ссылки и кликов по ней Создана ссылка "tensor" в body 1`] = `
<body>
  <a
    href="https://tensor.ru/"
  >
    tensor
  </a>
</body>
`;

exports[`Урок 5.17 - События 4. Создание списка, кнопки и кликов по ним Клик по кнопке 1`] = `
<body>
  <ul>
    <li>
      Пункт
    </li>
    <li>
      Пункт
    </li>
  </ul>
  <button>
    Добавить пункт
  </button>
</body>
`;

exports[`Урок 5.17 - События 4. Создание списка, кнопки и кликов по ним Клик по созданному по кнопке элементу списка 1`] = `
<body>
  <ul>
    <li>
      Пункт
    </li>
    <li>
      Пункт!
    </li>
  </ul>
  <button>
    Добавить пункт
  </button>
</body>
`;

exports[`Урок 5.17 - События 4. Создание списка, кнопки и кликов по ним Клик по элементу списка 1`] = `
<body>
  <ul>
    <li>
      Пункт!
    </li>
  </ul>
  <button>
    Добавить пункт
  </button>
</body>
`;

exports[`Урок 5.17 - События 4. Создание списка, кнопки и кликов по ним Создан список и кнопка 1`] = `
<body>
  <ul>
    <li>
      Пункт
    </li>
  </ul>
  <button>
    Добавить пункт
  </button>
</body>
`;
